### PR TITLE
Add solutions for chapter 2

### PR DESCRIPTION
**Pull request by jborkowski**
### Solutions for Chapter 2

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter2.hs [hunk 1]
```diff
@@ -336,7 +336,11 @@ from it!
 ghci> :l src/Chapter2.hs
 -}
 subList :: Int -> Int -> [a] -> [a]
-subList = error "subList: Not implemented!"
+subList from to ls
+  | from < 0 || to < 0 = []
+  | otherwise = take n . drop from $ ls
```
Comment (chshersh) [inline src/Chapter2.hs:341]: Your implementation is correct when the right bound is lower than the left. But I think you can implement it more efficiently by adding a separate case to your check :slightly_smiling_face:

File: src/Chapter2.hs [hunk 2]
```diff
@@ -348,8 +352,10 @@ Implement a function that returns only the first half of a given list.
 >>> firstHalf "bca"
 "b"
 -}
--- PUT THE FUNCTION TYPE IN HERE
-firstHalf l = error "firstHalf: Not implemented!"
+firstHalf :: [a] -> [a]
+firstHalf ls = fst . splitAt half $ ls
+  where
+    half = length ls `div` 2
```
Comment (chshersh) [inline src/Chapter2.hs:358]: This is a correct solution! Though, `splitAt` does more than required. Here, it's enough to _take_ the first half of the list

File: src/Chapter2.hs [hunk 3]
```diff
@@ -500,7 +506,10 @@ True
 >>> isThird42 [42, 42, 0, 42]
 False
 -}
-isThird42 = error "isThird42: Not implemented!"
+isThird42 :: (Eq a, Num a) => [a] -> Bool
+isThird42 []         = False
```
Comment (chshersh) [inline src/Chapter2.hs:510]: This case is redundant, since it will be covered by the `_` pattern on the last line

File: src/Chapter2.hs [hunk 4]
```diff
@@ -620,7 +629,10 @@ Write a function that takes elements of a list only on even positions.
 >>> takeEven [2, 1, 3, 5, 4]
 [2,3,4]
 -}
-takeEven = error "takeEven: Not implemented!"
+takeEven :: (Eq a, Num a) => [a] -> [a]
```
Comment (chshersh) [inline src/Chapter2.hs:632]: Constraints here are redundant. Notice that you don't use any specific information about values of type `a`, so no need to require anything special here :slightly_smiling_face: 
```suggestion
takeEven :: [a] -> [a]
```

File: src/Chapter2.hs [hunk 5]
```diff
@@ -740,8 +752,8 @@ the list with only those lists that contain a passed element.
 
 🕯 HINT: Use the 'elem' function to check whether an element belongs to a list
 -}
-contains = error "contains: Not implemented!"
-
+contains :: (Eq a, Num a) => a -> [[a]] -> [[a]]
```
Comment (chshersh) [inline src/Chapter2.hs:755]: The `Num` constraint is redundant here. Note, that `elem` needs only `Eq` for this function to work. We don't require people to write very polymorphic functions here, so by saying _list of lists of numbers_, we mean that a type signature like `Int -> [[Int]] -> [[Int]]` will be enough for this task. Nevertheless, nice work on specifying a polymorphic type signature :+1: 
```suggestion
contains :: (Eq a) => a -> [[a]] -> [[a]]
```

File: src/Chapter2.hs [hunk 6]
```diff
@@ -780,13 +792,14 @@ Let's now try to eta-reduce some of the functions and ensure that we
...
 
--- TODO: type ;)
-listElementsLessThan x l = filter (< x) l
+listElementsLessThan :: (Ord a) => a -> [a] -> [a]
+listElementsLessThan x = filter (< x)
 
 -- Can you eta-reduce this one???
-pairMul xs ys = zipWith (*) xs ys
+pairMul :: (Num a) => [a] -> [a] -> [a]
+pairMul = zipWith (*)
```
Comment (chshersh) [inline src/Chapter2.hs:802]: Eta-reduction award! :trophy: 

And nice work on general types :+1:

File: src/Chapter2.hs [hunk 7]
```diff
@@ -841,7 +854,12 @@ list.
 
 🕯 HINT: Use the 'cycle' function
 -}
-rotate = error "rotate: Not implemented!"
+rotate :: (Num a) => Int -> [a] -> [a]
+rotate n ls
+  | n < 0 = []
+  | otherwise = take listLen . drop n . cycle $ ls
+  where
+    listLen = length ls
```
Comment (chshersh) [inline src/Chapter2.hs:862]: Your implementation is correct. Though, you can optimize it even further. Note, that if `n` is a big number, `drop` will do a lot of work, if the list size is much smaller than the number `n`. It looks like it possible to optimize this function a bit :slightly_smiling_face:

File: src/Chapter2.hs [hunk 8]
```diff
@@ -857,7 +875,9 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+rewind :: [a] -> [a]
+rewind []    = []
+rewind (h:t) = rewind t ++ [h]
```
Comment (chshersh) [inline src/Chapter2.hs:880]: Your solution is correct! However, it is slow. In lists, it is quite slow to add anything at the end of the list. That is why it is always better to rewrite it with the `:` cons. Remember the explanation with trains? 🚂 🚋 🚋 

That is why a more efficient solution is with the accumulator and the recursive function that will do the addition at the start of the list which is instant!

You can read a bit more about the `go` pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#recursive-go

**Review by chshersh — APPROVED** [review]
Great work on Chapter Two tasks!
And well done on the Advanced Task :clap: 

I left some notes in my review. Keep up the great work :muscle:

**Comment by jborkowski** [comment]
@chshersh Thanks for comments! I improved my code :)

**Comment by chshersh** [comment]
@jborkowski Nice work! The solutions look good :+1: I'm going to close this PR, because we can't merge the solutions to our repo :slightly_smiling_face: But don't worry, it still counts towards Hacktoberfest contribution :shirt: :deciduous_tree: Feel free to open PRs with solutions to other chapters!

**Pull request closed without merging**